File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
import java.util.*;
class Solution {
    public int solution(int[] mats, String[][] park) {
        int answer = -1;
        int maxSize = 0;
        int[][] dp = new int[park.length][park[0].length];
        
        for(int i=0; i<park.length; i++) {
            for(int j=0; j<park[0].length; j++) {
                if(park[i][j].equals("-1")) {
                    if(i==0 || j==0) {
                        dp[i][j] = 1;
                    } else {
                        dp[i][j] = Math.min(Math.min(dp[i][j-1], dp[i-1][j]), dp[i-1][j-1]) + 1;
                    }
                    maxSize = Math.max(maxSize, dp[i][j]);
                }
            }
        }
        
        Arrays.sort(mats);
        for(int i=mats.length-1; i>=0; i--) {
            if(mats[i] <= maxSize) return mats[i];
        }
        
        return answer;
    }
}

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    answer = -1
    
    for i in range(len(park)) :
        for j in range(len(park[i])) :
            if park[i][j] == "-1" :
                for n in mats :
                    if isPossible(i,j,n,park) :
                        answer = max(answer, n)
    
    return answer

def isPossible(r, c, size, park) :
    for i in range(size) :
        for j in range(size) :
            nr = r+i
            nc = c+j
            
            if nr>=len(park) or nc>=len(park[0]) or park[nr][nc] != "-1" :
                return False
    return True